### PR TITLE
feat(map_based_prediction): enable prediction to/from shoulder lanelet (lanelet not in routing graph)

### PR DESCRIPTION
**Pull request by YoshiRi**
## Description

This PR is aiming to generate prediction path from/to a shoulder lanelet, which is unconnected in routing graph.

### Approaches

Implemented algorithm has following architecture:

- left/right shoulder lane prediction(center path prediction has similar architecture)
image

#### shoulder lane detection

- when two lanelet shares left/right linestring, we judge these lanelets have neighbor relationships

#### lanelet isolation detection

- a lanelet is considered as "isolated" when it does not have any left/right/following lanelet

#### lanelet length validation

- when lanelet length is not enough, predicted path could be oddly curved
- validation tries to evaluate the lanelet has enough lane length
  - minimum lanelet length is calculated by `obj linear velocity * time_horizon * rate`
  - In default, time_horizon is `10` [s] and rate is `0.8`

## Related links

- [ ] [PASS TIER IV INTERNAL SCENARIOS](https://evaluation.tier4.jp/evaluation/reports/b4321154-eb72-584a-9194-53e81e8156d4?project_id=prd_jt)

## Tests performed

Tested in psim.
I'll show edge cases:

| cases |result image | 
|--|--|
| normal lane to shoulder|Screenshot from 2023-07-21 17-25-12|
|shoulder lane to normal|Screenshot from 2023-07-21 18-12-24|
| normal lane to shoulder <br> (path discarded due to not enough lanelet length)|image|
|shoulder lane to normal <br> (path discarded due to not enough lanelet length)|Screenshot from 2023-07-21 18-12-50|

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: perception/map_based_prediction/src/map_based_prediction_node.cpp [hunk 1]
```diff
@@ -1119,25 +1231,84 @@ std::vector<PredictedRefPath> MapBasedPredictionNode::getPredictedReferencePath(
...
+            continue;
+          // get possible path for isolated lanelet
+          left_paths = getPossiblePathsForIsolatedLanelet(unconnected_left_lanelet);
+        } else {
+          // if it has connection it should have paths
+          left_paths = routing_graph_ptr_->possiblePaths(unconnected_left_lanelet, possible_params);
+          if (!left_paths.empty()) break;
+        }
+      }
     }
```
Comment (kyoichi-sugahara) [inline perception/map_based_prediction/src/map_based_prediction_node.cpp:1266]: just a suggestion you can ignore

consolidate implementation because it seems getting right and left lanelets are the same, 
if you use the lammda, the code can be like this?
```    
    // Step1.1 and Step1.2 Get the lanelets
    auto getLaneletPaths = [&](auto get_direction, auto get_line_sharing, auto maneuver) {
      lanelet::routing::LaneletPaths paths;
      auto opt_dir = get_direction(current_lanelet_data.lanelet);
      auto adjacent_dir = get_direction == routing_graph_ptr_->right
                            ? routing_graph_ptr_->adjacentRight(current_lanelet_data.lanelet)
                            : routing_graph_ptr_->adjacentLeft(current_lanelet_data.lanelet);
      if (!!opt_dir) {
        paths = routing_graph_ptr_->possiblePaths(*opt_dir, possible_params);
      } else if (!!adjacent_dir) {
        paths = routing_graph_ptr_->possiblePaths(*adjacent_dir, possible_params);
      } else {
        auto unconnected_dir = get_line_sharing(current_lanelet_data.lanelet, lanelet_map_ptr_);
        for (const auto & unconnected_lanelet : unconnected_dir) {
          if (isIsolatedLanelet(unconnected_lanelet, routing_graph_ptr_)) {
            if (!validateIsolatedLaneletLength(unconnected_lanelet, object, validate_time_horizon))
              continue;
            paths = getPossiblePathsForIsolatedLanelet(unconnected_lanelet);
          } else {
            paths = routing_graph_ptr_->possiblePaths(unconnected_lanelet, possible_params);
            if (!paths.empty()) break;
          }
        }
      }
      return paths;
    };

    lanelet::routing::LaneletPaths left_paths = getLaneletPaths(
      routing_graph_ptr_->left, getLeftLineSharingLanelets, Maneuver::LEFT_LANE_CHANGE);
    lanelet::routing::LaneletPaths right_paths = getLaneletPaths(
      routing_graph_ptr_->right, getRightLineSharingLanelets, Maneuver::RIGHT_LANE_CHANGE);
```

File: perception/map_based_prediction/src/map_based_prediction_node.cpp [hunk 2]
```diff
@@ -1119,25 +1231,84 @@ std::vector<PredictedRefPath> MapBasedPredictionNode::getPredictedReferencePath(
...
 
     // Step1.2 Get the right lanelet
     lanelet::routing::LaneletPaths right_paths;
     auto opt_right = routing_graph_ptr_->right(current_lanelet_data.lanelet);
+    auto adjacent_right = routing_graph_ptr_->adjacentRight(current_lanelet_data.lanelet);
     if (!!opt_right) {
       right_paths = routing_graph_ptr_->possiblePaths(*opt_right, possible_params);
+    } else if (!!adjacent_right) {
+      std::cout << adjacent_right->id() << std::endl;
```
[the remark this answers is not in the document]
Comment (YoshiRi) [reply perception/map_based_prediction/src/map_based_prediction_node.cpp]: Yes, this is fixed in https://github.com/autowarefoundation/autoware.universe/pull/4353/commits/953e66c825bdd9fac3227fe8a2000eed45d7bf77.

File: perception/map_based_prediction/src/map_based_prediction_node.cpp [hunk 3]
```diff
@@ -1119,25 +1231,84 @@ std::vector<PredictedRefPath> MapBasedPredictionNode::getPredictedReferencePath(
...
+      right_paths = routing_graph_ptr_->possiblePaths(*adjacent_right, possible_params);
+    } else {
+      // if no right lanele in graph
```
[the remark this answers is not in the document]
Comment (YoshiRi) [reply perception/map_based_prediction/src/map_based_prediction_node.cpp]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/4353/commits/953e66c825bdd9fac3227fe8a2000eed45d7bf77.

File: perception/map_based_prediction/src/map_based_prediction_node.cpp (hunk 1 above)
Comment (YoshiRi) [reply perception/map_based_prediction/src/map_based_prediction_node.cpp:1286]: Your suggestion make sense.
I think it is better to separate neighbor lanelets detection process and paths creation process, so added following changes:

- https://github.com/autowarefoundation/autoware.universe/pull/4353/commits/e3762cc4e2208963085a9e769892f0041be04240 : paths estimation to lambda function
- https://github.com/autowarefoundation/autoware.universe/pull/4353/commits/b69b3b38e85673199d7f83fc12535f1a4d7b3e4a : lanelet extraction process to lambda function

**Review by takayuki5168 — APPROVED** [review]

**Review by purewater0901 — APPROVED** [review]
